crates/turbopack-node/js/src/ipc/evaluate.ts

```diff
@@ -1,64 +1,135 @@
 import { IPC, StructuredError } from "./index";
 import type { Ipc as GenericIpc } from "./index";
 
-type IpcIncomingMessage = {
-  type: "evaluate";
-  args: string[];
-};
-
-type IpcOutgoingMessage =
+type IpcIncomingMessage =
   | {
-      type: "value";
-      data: string;
+      type: "evaluate";
+      args: string[];
     }
+  | {
+      type: "result";
+      id: number;
+      error: string | null;
+      data: any | null;
+    };
+
+type IpcOutgoingMessage =
   | {
       type: "end";
       data: string | undefined;
+      duration: number;
     }
   | {
-      type: "fileDependency";
-      path: string;
-    }
-  | {
-      type: "buildDependency";
-      path: string;
+      type: "info";
+      data: any;
     }
   | {
-      type: "dirDependency";
-      path: string;
-      glob: string;
-    }
-  | {
-      type: "emittedError";
-      severity: "warning" | "error";
-      error: StructuredError;
+      type: "request";
+      id: number;
+      data: any;
     };
 
-export type Ipc = GenericIpc<IpcIncomingMessage, IpcOutgoingMessage>;
-const ipc = IPC as Ipc;
+export type Ipc<IM, RM> = {
+  sendInfo(message: IM): Promise<void>;
+  sendRequest(message: RM): Promise<unknown>;
+  sendError(error: Error): Promise<never>;
+};
+const ipc = IPC as GenericIpc<IpcIncomingMessage, IpcOutgoingMessage>;
+
+const queue: string[][] = [];
 
 export const run = async (
-  getValue: (ipc: Ipc, ...deserializedArgs: any[]) => any
+  moduleFactory: () => Promise<{
+    init?: () => Promise<void>;
+    default: (ipc: Ipc<any, any>, ...deserializedArgs: any[]) => any;
+  }>
 ) => {
+  let nextId = 1;
+  const requests = new Map();
+  const internalIpc = {
+    sendInfo: (message: any) =>
+      ipc.send({
+        type: "info",
+        data: message,
+      }),
+    sendRequest: (message: any) => {
+      const id = nextId++;
+      let resolve, reject;
+      const promise = new Promise((res, rej) => {
+        resolve = res;
+        reject = rej;
+      });
+      requests.set(id, { resolve, reject });
+      return ipc
+        .send({ type: "request", id, data: message })
+        .then(() => promise);
+    },
+    sendError: (error: Error) => {
+      return ipc.sendError(error);
+    },
+  };
+
+  // Initialize module and send ready message
+  let getValue: (ipc: Ipc<any, any>, ...deserializedArgs: any[]) => any;
+  try {
+    const module = await moduleFactory();
+    if (typeof module.init === "function") {
+      await module.init();
+    }
+    getValue = module.default;
+  } catch (err) {
+    await ipc.sendReady();
+    await ipc.sendError(err as Error);
+  }
+  await ipc.sendReady();
```

Comment:
Seems this will send ready twice, should it look like this:
```suggestion
   await ipc.sendReady();
  } catch (err) {
    await ipc.sendReady();
    await ipc.sendError(err as Error);
  }

```